assignments/serde-lifetimes.adoc

```diff
@@ -0,0 +1,148 @@
+= Exercise: Copy-on-Write Data with Serde
+:source-language: rust
+
+In this exercise you will learn
+
+* How to serialize and deserialize data with Serde crate.
+* When to use types that own their data or only refer to externally owned data
+* How to use `std::borrow::Cow` type to minimize extra memory allocations
+
+== Getting Started
+
+Create a new binary project for this exercise. Add the following dependencies to your `Cargo.toml` file:
+
+[source,toml]
+----
+[dependencies]
+serde = { version = "1", features = ["derive"] }
+serde_json = "1"
+----
+
+Use this template:
+[source,rust]
+----
+// pretend that we call an API and get a JSON String back
+fn fetch_data() -> String {
+    String::from(
+        r#"
+            {
+                "id": 1,
+                "title": "Hello, Rust"
+            }
+        "#,
+    )
+}
+
+#[derive(Debug)]
+struct BlogPost {
+    id: u32,
+
+    title: String,
+}
+
+fn main() -> Result<(), serde_json::Error> {
+    let post: BlogPost = {
+        let data = fetch_data();
+        todo!("use `serde_json` crate to parse JSON")
+    };
+    println!("deserialized = {:?}", post);
+
+    let post_json: String =
+        todo!("use `serde_json` to convert `post` to a string");
+    println!("serialized = {:?}", post_json);
+
+    Ok(())
+}
+----
+
+.What is that `r#"...` thing?
+[%collapsible]
+====
+`r` in front of a string literal means it's a "raw" string. Escape sequences (`\n`, `\"`, etc.) don't work, and thus they are very convenient for things like regular expressions, JSON literals, etc.
+
+Optionally `r` can be followed by one or more symbols (like `\#` in our case), and then your string ends when there's a closing double quote followed by the same number of the same symbols. This is great for cases when you want to have double quotes inside your string literal. For our example `r#" ... "#` works great for JSON. In rare cases you'd want to put two or more pound signs. Like, when you store CSS color values in your JSON strings:
+[source,rust]
+----
+// here `"#` would not terminate the string
+r##"
+    {
+        "color": "#ff00ff"
+    }
+"##
+----
+====
+
+== Tasks
+
+=== 1. Serialization and Deserialization using `String` s
```

Comment:
Perhaps it is a good idea to mention that you want the participants to do all `todo`s here in the Task 1 itself.

Reply:
Isn't it what the very next line says?

Reply:
Ah yea, I think I got confused that should this step one should do all `todo()`s or not. I guess its not needed. My bad.